ion-schema/src/isl/isl_type_reference.rs

```diff
@@ -311,3 +326,104 @@ impl IslTypeRefImpl {
         }
     }
 }
+
+/// Represents a [variably occurring type reference] that will be used by `ordered_elements` and `fields` constraints.
+///  
+/// ```ion
+/// Grammar: <VARIABLY_OCCURRING_TYPE_ARGUMENT> ::= { <OCCURS>, <CONSTRAINT>... }
+///                                      | <TYPE_ARGUMENT>
+///
+///         <OCCURS> ::= occurs: <INT>
+///            | occurs: <RANGE_INT>
+///            | occurs: optional
+///            | occurs: required
+/// ```
+///
+/// [variably occurring type reference]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#variably-occurring-type-arguments
+#[derive(Debug, Clone, PartialEq)]
+pub struct IslVariablyOccurringTypeRef {
+    type_ref: IslTypeRefImpl,
+    occurs: Range,
+}
+
+impl IslVariablyOccurringTypeRef {
+    pub(crate) fn new(type_ref: IslTypeRef, occurs: Range) -> Self {
+        Self {
+            type_ref: type_ref.type_reference,
+            occurs,
+        }
+    }
+
+    pub fn optional(type_ref: IslTypeRef) -> Self {
+        Self {
+            type_ref: type_ref.type_reference,
+            occurs: Range::optional(),
+        }
+    }
+
+    pub fn required(type_ref: IslTypeRef) -> Self {
+        Self {
+            type_ref: type_ref.type_reference,
+            occurs: Range::required(),
+        }
+    }
+
+    pub fn occurs(&self) -> Range {
+        self.occurs.to_owned()
+    }
+
+    /// Tries to create an [IslVariablyOccurringTypeRef] from the given Element
+    pub fn from_ion_element(
+        constraint_name: &str,
+        isl_version: IslVersion,
+        value: &Element,
+        inline_imported_types: &mut Vec<IslImportType>,
+    ) -> IonSchemaResult<Self> {
+        let type_ref = IslTypeRefImpl::from_ion_element_with_occurs_flag(
+            isl_version,
+            value,
+            inline_imported_types,
+            true,
+        )?;
+
+        let occurs: Range = if let IslTypeRefImpl::Anonymous(isl_type, _) = &type_ref {
```

Comment:
If you change this line to `let occurs: Option<Range> = ...`, then you can eliminate most/all of L397-408, and replace the return value with
```rust
match occurs {
    Some(range) => Ok(IslVariablyOccurringTypeRef::new(type_ref, occurs)),
    None if constraint_name == "fields" => Ok(IslVariablyOccurringTypeRef::optional(type_ref)),
    None => Ok(IslVariablyOccurringTypeRef::required(type_ref))
}
```

That is one possible way to fix the clippy warning and to get rid of the duplicated if/else expression.

Reply:
I have used a similar solution using `and_then` and `unwrap_or`. Please let me know if that sounds good?

Reply:
Yes. Looks good.